### PR TITLE
🎉 MySQL destination: normalization

### LINKED ISSUE #3429: Implement normalization for the MySql destination.
## Tell us about the problem you're trying to solve
MySql currently does not support normalization. It should support it.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368197699942) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by tuliren**
## What
- This PR resolves #3429.

## How
*Describe the solution*

## Todo
- [ ] ~Append hash to table name~
    - I could not fix all the test failures after making this change on both Java and Python
    - This change would affect existing normalizations and requires a migration
- [x] ~Investigate dependency specification~
    - Move `dbt-mysql` to `Dockerfile` does not solve the problem
        - `pip install dbt-mysql` there still resulted in missing dependency exception related to `dbt`.
   - Cannot change `dbt` version to 0.19.1
       - `dbt-mysql` requires `dbt` 0.19.0
       - Forking our own `dbt-mysql` that depends on `dbt` 0.19.1 resulted in a `charset` issue and test failures.
           - This approach requires more investigation.
- [x] Add unit test for mysql name converter
- [x] Check in mysql dbt output
- [x] Update changed dbt output

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<summary> <strong> Connector checklist </strong> </summary>
<p>

- [x] Issue acceptance criteria met
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [x] Secrets are annotated with `airbyte_secret` in output spec
- [x] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [x] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [x] Documentation updated 
    - [x] README
    - [x] CHANGELOG.md
    - [x] Reference docs in the `docs/integrations/` directory.
    - [x] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [x] Build is successful
- [x] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [x] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>

### CODE REVIEW

**Comment by tuliren** [comment]
@ChristopheDuong, I am running into this error in the MySQL integration test:

```
java.sql.SQLSyntaxErrorException: Identifier name '_airbyte_raw_nested_stream_with_complex_columns_resulting_into_long_names' is too long
```

Do you know why the truncation is not working? It worked in the unit tests.

Here is the [logs](https://github.com/airbytehq/airbyte/runs/2860941346?check_suite_focus=true).

**Comment by ChristopheDuong** [comment]
> Do you know why the truncation is not working? It worked in the unit tests.
It seems to be working on table names: `nested_stream_with_co__lting_into_long_names_ab1`
but not applied to column names: `_airbyte_raw_nested_stream_with_complex_columns_resulting_into_long_names` from the logs...

**Comment by ChristopheDuong** [comment]
> Do you know how would `destination-mysql` choose to call the output raw table for such a stream?

After downloading and looking at the output log from the `destination-mysql`, here is the answer:

```
2021-06-18 18:49:56 [1;31mERROR[m i.a.i.d.b.BufferedStreamConsumer(close):203 - {} - on close failed.
java.sql.SQLSyntaxErrorException: Identifier name '_airbyte_raw_nested_stream_with_complex_columns_resulting_into_long_names' is too long
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:120) ~[mysql-connector-java-8.0.22.jar:8.0.22]
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97) ~[mysql-connector-java-8.0.22.jar:8.0.22]
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122) ~[mysql-connector-java-8.0.22.jar:8.0.22]
	at com.mysql.cj.jdbc.StatementImpl.executeInternal(StatementImpl.java:764) ~[mysql-connector-java-8.0.22.jar:8.0.22]
	at com.mysql.cj.jdbc.StatementImpl.execute(StatementImpl.java:648) ~[mysql-connector-java-8.0.22.jar:8.0.22]
	at org.apache.commons.dbcp2.DelegatingStatement.execute(DelegatingStatement.java:194) ~[commons-dbcp2-2.7.0.jar:2.7.0]
	at org.apache.commons.dbcp2.DelegatingStatement.execute(DelegatingStatement.java:194) ~[commons-dbcp2-2.7.0.jar:2.7.0]
	at io.airbyte.db.jdbc.JdbcDatabase.lambda$execute$0(JdbcDatabase.java:50) ~[io.airbyte-airbyte-db-0.26.0-alpha.jar:?]
	at io.airbyte.db.jdbc.DefaultJdbcDatabase.execute(DefaultJdbcDatabase.java:61) ~[io.airbyte-airbyte-db-0.26.0-alpha.jar:?]
	at io.airbyte.db.jdbc.JdbcDatabase.execute(JdbcDatabase.java:50) ~[io.airbyte-airbyte-db-0.26.0-alpha.jar:?]
	at io.airbyte.integrations.destination.jdbc.DefaultSqlOperations.createTableIfNotExists(DefaultSqlOperations.java:65) ~[io.airbyte.airbyte-integrations.connectors-destination-jdbc-0.26.0-alpha.jar:?]
	at io.airbyte.integrations.destination.jdbc.JdbcBufferedConsumerFactory.lambda$onCloseFunction$3(JdbcBufferedConsumerFactory.java:190) ~[io.airbyte.airbyte-integrations.connectors-destination-jdbc-0.26.0-alpha.jar:?]
	at io.airbyte.integrations.destination.buffered_stream_consumer.OnCloseFunction.accept(OnCloseFunction.java:29) ~[io.airbyte.airbyte-integrations.bases-base-java-0.26.0-alpha.jar:?]
	at io.airbyte.integrations.destination.buffered_stream_consumer.BufferedStreamConsumer.close(BufferedStreamConsumer.java:195) [io.airbyte.airbyte-integrations.bases-base-java-0.26.0-alpha.jar:?]
	at io.airbyte.integrations.base.FailureTrackingAirbyteMessageConsumer.close(FailureTrackingAirbyteMessageConsumer.java:82) [io.airbyte.airbyte-integrations.bases-base-java-0.26.0-alpha.jar:?]
	at io.airbyte.integrations.base.IntegrationRunner.consumeWriteStream(IntegrationRunner.java:138) [io.airbyte.airbyte-integrations.bases-base-java-0.26.0-alpha.jar:?]
	at io.airbyte.integrations.base.IntegrationRunner.run(IntegrationRunner.java:113) [io.airbyte.airbyte-integrations.bases-base-java-0.26.0-alpha.jar:?]
	at io.airbyte.integrations.destination.mysql.MySQLDestination.main(MySQLDestination.java:111) [io.airbyte.airbyte-integrations.connectors-destination-mysql-0.26.0-alpha.jar:?]
```

File: airbyte-integrations/bases/base-normalization/unit_tests/test_table_name_registry.py [hunk 1]
```diff
@@ -129,26 +129,29 @@ def read_json(input_path: str, apply_function=(lambda x: x)):
 
 
 @pytest.mark.parametrize(
-    "json_path, expected_postgres, expected_bigquery",
+    "json_path, expected_postgres, expected_bigquery, expected_mysql",
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/unit_tests/test_table_name_registry.py:132]: This test is not intended to be exhaustive over the destinations, so it's not mandatory to add your new destination expected field here.

The original intent here was to only unit test simple_name vs nested_hashed_name functions in the table_name_registry

File: airbyte-integrations/bases/base-normalization/integration_tests/resources/test_primary_key_streams/dbt_schema_tests/schema_test.yml [hunk 2]
```diff
@@ -4,11 +4,11 @@ models:
...
-          description: check_streams_are_equal
-            In this integration test, we are sending the same records to both streams
-            exchange_rate and dedup_exchange_rate.
-            The SCD table of dedup_exchange_rate in append_dedup mode should therefore mirror
-            the final table with append or overwrite mode from exchange_rate.
+          # description: check_streams_are_equal
+          #  In this integration test, we are sending the same records to both streams
+          #  exchange_rate and dedup_exchange_rate.
+          #  The SCD table of dedup_exchange_rate in append_dedup mode should therefore mirror
+          #  the final table with append or overwrite mode from exchange_rate.
```
Comment (tuliren) [inline airbyte-integrations/bases/base-normalization/integration_tests/resources/test_primary_key_streams/dbt_schema_tests/schema_test.yml:11]: If I don't comment the `description` out, DBT will throw the following error:

```
macro 'dbt_macro__mysql__test_equality' takes no keyword argument 'description'
```
which makes sense, because the `description` is not a parameter for this macro. But I am confused why this error is not throw for other DB or tests.

Comment (ChristopheDuong) [reply airbyte-integrations/bases/base-normalization/integration_tests/resources/test_primary_key_streams/dbt_schema_tests/schema_test.yml:11]: Weird that it's breaking only now...

 I've been putting these descriptions fields for documenting our tests. Commenting them out are fine too though...

**Comment by tuliren** [comment]
The issues are:

One warning about missing identifier name, even after I have truncated the table names:
```
[WARNING]: Test 'test.airbyte_utils.test_check_row_counts' (tests/test_check_row_counts.sql) depends on a source named 'test_normalization._airbyte_raw_nested_stream_with_complex_columns_resulting_into_long_names' which was not found
```

2 Errors:

```
Compilation Error in test dbt_utils_equality_exchange_rate_id__currency__date___HKD_sp_i_l_characters___HKD_special___characters__NZD__USD__ref_dedup_exchange_rate_scd_ (models/dbt_schema_tests/schema_test.yml)
```

```
8 of 8 FAIL 2 not_null_nested_stream_with_co___names_partition_data__currency_............................... [FAIL 2 in 4.47s]

Bad test equality: Returned 0 rows and 1 cols, but expected 1 row and 1 column
```

**Comment by ChristopheDuong** [comment]
> The issues are:
> 
> One warning about missing identifier name, even after I have truncated the table names:
> 
> ```
> [WARNING]: Test 'test.airbyte_utils.test_check_row_counts' (tests/test_check_row_counts.sql) depends on a source named 'test_normalization._airbyte_raw_nested_stream_with_complex_columns_resulting_into_long_names' which was not found
> ```
> 
> 2 Errors:
> 
> ```
> Compilation Error in test dbt_utils_equality_exchange_rate_id__currency__date___HKD_sp_i_l_characters___HKD_special___characters__NZD__USD__ref_dedup_exchange_rate_scd_ (models/dbt_schema_tests/schema_test.yml)
> ```
> 
> ```
> 8 of 8 FAIL 2 not_null_nested_stream_with_co___names_partition_data__currency_............................... [FAIL 2 in 4.47s]
> 
> Bad test equality: Returned 0 rows and 1 cols, but expected 1 row and 1 column
> ```

Yes, you renamed it in the normalization code that generates the dbt models... but the errors is in the tests files that were configured with the old names (these are not regenerated by the code and needs to be adapted accordingly)

**Comment by ChristopheDuong** [comment]
> 8 of 8 FAIL 2 not_null_nested_stream_with_co___names_partition_data__currency_............................... [FAIL 2 in 4.47s]
> Bad test equality: Returned 0 rows and 1 cols, but expected 1 row and 1 column

This was expecting a row with a value in this column but it didn't... is it because the unnesting didn't work in MySQL?

**Comment by ChristopheDuong** [comment]
I'm unable to run the tests locally either...

Have you added the MySQL-adapter git repo to `airbyte-integrations/bases/base-normalization/dbt-project-template/packages.yml` btw?

**Comment by tuliren** [comment]
> > 8 of 8 FAIL 2 not_null_nested_stream_with_co___names_partition_data__currency_............................... [FAIL 2 in 4.47s]
> > Bad test equality: Returned 0 rows and 1 cols, but expected 1 row and 1 column
> 
> This was expecting a row with a value in this column but it didn't... is it because the unnesting didn't work in MySQL?

Unnesting worked for other tests, like json array. So it should work for this test case as well.

I need to dig more into this failure.

File: airbyte-integrations/bases/base-normalization/setup.py [hunk 3]
```diff
@@ -32,7 +32,14 @@
...
     packages=setuptools.find_packages(),
-    install_requires=["airbyte-protocol", "pyyaml", "jinja2"],
+    install_requires=[
+        "airbyte-protocol",
+        "pyyaml",
+        "jinja2",
+        # TODO: change to dbt-mysql after this PR is merged:
+        # https://github.com/dbeatty10/dbt-mysql/pull/57
+        "dbt-mysql @ git+https://github.com/airbytehq/dbt-mysql-1@patch-1#egg=dbt-mysql",
+    ],
```
Comment (tuliren) [inline airbyte-integrations/bases/base-normalization/setup.py:42]: > I'm unable to run the tests locally either...
> 
> Have you added the MySQL-adapter git repo to `airbyte-integrations/bases/base-normalization/dbt-project-template/packages.yml` btw?

@ChristopheDuong, I did not add it to the `packages.yml` but added it here. It was necessary because when I started working on this PR, there was a bug in the `dbt-mysql` release, and a [patch](https://github.com/dbeatty10/dbt-mysql/pull/57) was needed. This patch just got merged 2 days ago. I will test and see if I can just add it to the dbt dependency instead of here.

**Comment by tuliren** [comment]
> > The issues are:
> > One warning about missing identifier name, even after I have truncated the table names:
> > ```
> > [WARNING]: Test 'test.airbyte_utils.test_check_row_counts' (tests/test_check_row_counts.sql) depends on a source named 'test_normalization._airbyte_raw_nested_stream_with_complex_columns_resulting_into_long_names' which was not found
> > ```
> 
> Yes, you renamed it in the normalization code that generates the dbt models... but the errors are in the tests files that are still configured manually with the old names (these are not regenerated by the code and needs to be adapted accordingly)
> 
> Thus, you should add rules to rename those long names in test files when used with MySQL such as `_airbyte_raw_nested_stream_with_complex_columns_resulting_into_long_names` to be replaced by adequate new value (handled by `airbyte-integrations/bases/base-normalization/integration_tests/resources/test_primary_key_streams/data_input/replace_identifiers.json`)

Got it. I already added an entry for `nested_stream_with_complex_columns_resulting_into_long_names` in `replace_identifiers.json`. I thought that was enough (without the `_airbyte_raw` prefix). But will try to add the full name there directly.

**Comment by tuliren** [comment]
Finally all tests have passed. I could not figure out the root cause earlier because I was looking at the `master` branch of `dbt-utils`, while we are using version 0.6.4, which is quite different from the `master` branch.

**Comment by tuliren** [comment]
Hmm, still one mysql failure, and many non-mysql failures:
```
FAILED integration_tests/test_ephemeral.py::test_destination_supported_limits[DestinationType.BIGQUERY-1500]
FAILED integration_tests/test_ephemeral.py::test_destination_supported_limits[DestinationType.REDSHIFT-1500]
FAILED integration_tests/test_ephemeral.py::test_destination_supported_limits[DestinationType.MYSQL-1500]
FAILED integration_tests/test_ephemeral.py::test_destination_failure_over_limits[BigQuery-2500-The view is too large.]
FAILED integration_tests/test_ephemeral.py::test_destination_failure_over_limits[Redshift-1665-target lists can have at most 1664 entries]
FAILED integration_tests/test_normalization.py::test_normalization[DestinationType.BIGQUERY-test_primary_key_streams]
FAILED integration_tests/test_normalization.py::test_normalization[DestinationType.REDSHIFT-test_primary_key_streams]
```

**Comment by ChristopheDuong** [comment]
> Finally all tests have passed. I could not figure out the root cause earlier because I was looking at the `master` branch of `dbt-utils`, while we are using version 0.6.4, which is quite different from the `master` branch.

If you can upgrade `dbt-utils` without encountering too many obstacles, go ahead!

if it makes it easier on MySQL, it's even better...

**Comment by tuliren** [comment]
All MySQL failures have been fixed. But there are failures from other destinations:

```
FAILED integration_tests/test_ephemeral.py::test_destination_supported_limits[DestinationType.BIGQUERY-1500]
FAILED integration_tests/test_ephemeral.py::test_destination_supported_limits[DestinationType.REDSHIFT-1500]
FAILED integration_tests/test_ephemeral.py::test_destination_failure_over_limits[BigQuery-2500-The view is too large.]
FAILED integration_tests/test_ephemeral.py::test_destination_failure_over_limits[Redshift-1665-target lists can have at most 1664 entries]
FAILED integration_tests/test_normalization.py::test_normalization[DestinationType.BIGQUERY-test_primary_key_streams]
FAILED integration_tests/test_normalization.py::test_normalization[DestinationType.REDSHIFT-test_primary_key_streams]
```

File: airbyte-integrations/bases/base-normalization/integration_tests/test_ephemeral.py [hunk 4]
```diff
@@ -61,8 +61,18 @@ def setup_test_path(request):
...
+@pytest.mark.parametrize(
+    "integration_type",
+    [
+        DestinationType.POSTGRES,
+        DestinationType.BIGQUERY,
+        DestinationType.SNOWFLAKE,
+        DestinationType.REDSHIFT,
+        # MySQL: the max number of columns is limited by row size (8KB),
+        # not absolute column count. It is way fewer than 1500.
+    ],
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/integration_tests/test_ephemeral.py:73]: This is not going over the full list of values of DestinationType enum anymore, so if we add a new destination, how do we make sure we don't forget to test it with maximum numbers of columns?

Would it be better to keep the parameter as `list(DestinationType)` but add an `if destination = MySQL then return` in the function code to skip the test for MySQL only?

File: airbyte-integrations/bases/base-normalization/setup.py (hunk 3 above)
Comment (tuliren) [reply airbyte-integrations/bases/base-normalization/setup.py:42]: Although the patch has been merged, however, we cannot add `dbt-mysql` under `packages.yml` because it has no `dbt_project.yml`:

```
no dbt_project.yml found
```

So we can only specify it in `setup.py`.

File: airbyte-integrations/bases/base-normalization/integration_tests/test_ephemeral.py (hunk 4 above)
Comment (tuliren) [reply airbyte-integrations/bases/base-normalization/integration_tests/test_ephemeral.py]: Good point. Done.

File: airbyte-integrations/bases/base-normalization/setup.py [hunk 5]
```diff
@@ -32,7 +32,13 @@
     author_email="contact@airbyte.io",
     url="https://github.com/airbytehq/airbyte",
     packages=setuptools.find_packages(),
-    install_requires=["airbyte-protocol", "pyyaml", "jinja2"],
+    install_requires=[
+        "airbyte-protocol",
+        "pyyaml",
+        "jinja2",
+        "dbt==0.19.0",
```
[the remark this answers is not in the document]
Comment (tuliren) [reply airbyte-integrations/bases/base-normalization/setup.py:39]: I am experimenting on removing it. Will try the docker idea.

**Comment by tuliren** [comment]
@ChristopheDuong, this PR is ready for review.

However, I was not able to do the following with this PR:
- Remove `dbt-mysql` and `dbt` dependencies from `setup`
    - I tried to move `dbt-mysql` to `Dockerfile`. But installing `dbt-mysql` using `pip` does not solve the problem. There were still missing dependency exceptions related to `dbt`.
   - I also could not change `dbt` version to 0.19.1.
       - This is because `dbt-mysql` requires `dbt` 0.19.0.
       - I tried to fork our own `dbt-mysql` that depends on `dbt` 0.19.1. But that ended up with a `charset` issue, which failed the tests. I think this approach should eventually work. But it requires more investigation. The root cause is probably due to some changes from `dbt` 0.19.0 to 0.19.1.
- Append a hash at the end of long table names
    - After making this change in both the Python and Java, I could not fix all the failed integration tests.
    - This change would affect existing normalizations and requires a migration, which seems beyong the scope of this PR.

Can we merge this PR as is for now? I can create issues to track the remaining problems.

I plan to check in the new mysql test output after the code review is done, because there are lots of files and can make it harder to review the code. But I push them if you prefer to review those files as well.

**Comment by ChristopheDuong** [comment]
It seems @marcosmarxm added an extra test case with the dedup fix with CDC...

It is so weird, MySQL runs this query fine:
Screenshot 2021-07-01 at 14 12 57

but fails with this because the column `_ab_cdc_deleted_at` contains `NULL` values.
Screenshot 2021-07-01 at 14 13 20

File: airbyte-integrations/bases/base-normalization/dbt-project-template/macros/cross_db_utils/datatypes.sql [hunk 6]
```diff
@@ -19,3 +19,11 @@
 {% macro snowflake__type_json() %}
     variant
 {% endmacro %}
+
+{%- macro mysql__type_json() -%}
+    json
+{%- endmacro -%}
+
+{%- macro mysql__type_string() -%}
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/dbt-project-template/macros/cross_db_utils/datatypes.sql:27]: maybe we should merge this file with the type conversions macros file?

File: airbyte-integrations/bases/base-normalization/dbt-project-template/macros/cross_db_utils/type_conversions.sql [hunk 7]
```diff
@@ -38,3 +38,8 @@
 {% macro redshift__cast_to_boolean(field) -%}
     cast(decode({{ field }}, 'true', '1', 'false', '0')::integer as boolean)
 {%- endmacro %}
+
+{# cast_to_bigint -------------------------------------------------     #}
+{% macro mysql__type_bigint() %}
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/dbt-project-template/macros/cross_db_utils/type_conversions.sql:43]: Should this go into `airbyte-integrations/bases/base-normalization/dbt-project-template/macros/cross_db_utils/datatypes.sql` instead?

this is extending the macro from dbt_utils right?

File: airbyte-integrations/bases/base-normalization/integration_tests/test_ephemeral.py [hunk 8]
```diff
@@ -62,7 +62,11 @@ def setup_test_path(request):
 
 @pytest.mark.parametrize("column_count", [1500])
 @pytest.mark.parametrize("integration_type", list(DestinationType))
-def test_destination_supported_limits(integration_type: DestinationType, column_count: int, setup_test_path):
+def test_destination_supported_limits(integration_type: DestinationType, column_count: int):
+    if integration_type == DestinationType.MYSQL:
+        # In MySQL, the max number of columns is limited by row size (8KB),
+        # not by absolute column count. It is way fewer than 1500.
+        return
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/integration_tests/test_ephemeral.py:69]: Not to be solved in this PR but we should probably handle the validity of rows (in terms of size) in the `destination-mysql` and log some warning/errors or reject such records?

About 250 columns, seems to be really low, especially for sources like hubspot/salesforce and they could easily run into exceptions there?

File: airbyte-integrations/bases/base-normalization/setup.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChristopheDuong) [reply airbyte-integrations/bases/base-normalization/setup.py:39]: does something like this work? i don't see the errors you were mentioning:
https://github.com/airbytehq/airbyte/pull/4459

**Comment by tuliren** [comment]
The root cause of the casting failure is that when using `json_extract`, the extracted json null is different from a native mysql null. The correct approach is to use `json_value` for extracting scaler json values, which will return native mysql types.

File: airbyte-integrations/bases/base-normalization/integration_tests/test_ephemeral.py (hunk 8 above)
Comment (tuliren) [reply airbyte-integrations/bases/base-normalization/integration_tests/test_ephemeral.py:69]: The max number of columns is related to the size of each column. In the test, we are creating string columns, which corresponds to `char` columns in mysql and each column is relatively large. That's why as few as 250 columns can fail the test. In reality, not every column will be of type `char`, and the max is 4096 in mysql.

**Review by ChristopheDuong — APPROVED** [review]

**Comment by ChristopheDuong** [comment]
good!

don't forget to bump and publish MySQL destination and the normalization image then
